marimo/_runtime/virtual_file.py

```diff
@@ -2,49 +2,94 @@
 from __future__ import annotations
 
 import dataclasses
-import io
-from typing import TYPE_CHECKING, Callable
+from multiprocessing import shared_memory
+from typing import TYPE_CHECKING
 
+from marimo import _loggers
 from marimo._runtime.cell_lifecycle_item import CellLifecycleItem
 
 if TYPE_CHECKING:
     from marimo._runtime.context import RuntimeContext
 
+LOGGER = _loggers.marimo_logger()
+
 
 @dataclasses.dataclass
 class VirtualFile:
     url: str
+    filename: str
+    buffer: bytes
 
-    def __init__(self, filename: str) -> None:
-        raise NotImplementedError
+    def __init__(self, filename: str, mimetype: str, buffer: bytes) -> None:
+        self.filename = filename
+        self.mimetype = mimetype
+        self.url = f"/@file/{filename}/{mimetype}"
```

Comment:
this produces something like `http://localhost:2718/@file/NiAlthcx.pdf/application/pdf`
vega (and possibly some other libraries) will look at the extension of the url to infer the mime.

thoughts on using the extension to convert to mime using `import mimetypes; mimetypes.guess_type`?